// @fanswoo/control-center/src/core/ajax-submit.ts

import { AjaxSubmit } from '@fanswoo/ajax-submit/main';
import { Message } from '@fanswoo/message/main';

export default class extends AjaxSubmit {
  Vue!: Vue | null;
  
  routePageStore: null;
  
  setStore(routePageStore) {
    this.routePageStore = routePageStore;
  }

  setVue(Vue: Vue) {
    this.Vue = Vue;
  }

  runDoneMessageShow(response) {
    const message = Message.getInstance();
    message.show({
      text: response.text,
      second: response.second,
      downloadUrl: response.downloadUrl,
    });

    if (response.syncORMs && response.syncORMs.ids && response.syncORMs.path) {
      this.routePageStore.commit('RoutePage/setSyncORMs', {
        ids: response.syncORMs.ids,
        path: response.syncORMs.path,
      });
    }

    if (response.syncORM && response.syncORM.path) {
      this.routePageStore.commit('RoutePage/setSyncORM', {
        id: response.syncORM.id,
        path: response.syncORM.path,
      });
    }

    if (response.fails) {
      return false;
    }

    if (response.rootUrl) {
      window.location.href = response.rootUrl;
      return true;
    }

    let lastPageUrl = '';
    let scrollTop = 0;
    
    if (
      this.routePageStore.state.RoutePage.lastPagePathHistory[
      this.routePageStore.state.RoutePage.lastPagePathHistory.length - 1
      ]
    ) {
      lastPageUrl =
        this.routePageStore.state.RoutePage.lastPagePathHistory[
        this.routePageStore.state.RoutePage.lastPagePathHistory.length - 1
        ].fullPath;
      scrollTop =
        this.routePageStore.state.RoutePage.lastPagePathHistory[
        this.routePageStore.state.RoutePage.lastPagePathHistory.length - 1
        ].scrollTop;
    }

    if (!response.url) {
      return false;
    }

    if (response.url === this.Vue?.$route.fullPath) {
      this.routePageStore.commit('RoutePage/resetPageData');
    } else if (response.url === lastPageUrl) {
      this.Vue?.$router.go(-1);
      $(window).scrollTop(scrollTop);
    } else {
      this.Vue?.$router.push(response.url);
    }

    return true;
  }
}
