// @fanswoo/package-manager/package-manager.js
module.exports = {
  platforms: [
    {
      name: 'composer',
      packages: [
        {
          name: 'fanswoo/framework-core',
          dev: false,
          sources: [
            {
              source: 'path',
              path: '../framework-core',
            },
            {
              source: 'github',
              name: 'fanswoo/framework-core',
              branch: 'develop',
            },
          ],
        },
      ],
    },
    {
      name: 'npm',
      dependenceNamespace: 'fanswoo-package-manager-dependence',
      dependenceDistDirectory: 'storage/framework/npm-dependence',
      packages: [
        {
          name: '@fanswoo/core',
          dev: false,
          sources: [
            {
              source: 'github',
              name: 'fanswoo/framework-core-front',
              branch: 'develop',
            },
            {
              source: 'path',
              path: 'storage/npm-packages/core',
            },
            {
              source: 'packagist',
              version: '^1.0',
            },
          ],
        },
      ],
    },
  ],
};
